pkg/database/redis_cache_test.go

```diff
@@ -0,0 +1,326 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/opencensus-integrations/redigo/redis"
+)
+
+type redisCommandInterceptorConn struct {
+	redis.ConnWithContext
+
+	mu   sync.RWMutex
+	seqs []*redisSequence
+}
+
+type redisSequence struct {
+	Kind    string
+	Command string
+	Args    []interface{}
+}
+
+func (rcic *redisCommandInterceptorConn) SendContext(ctx context.Context, commandName string, args ...interface{}) error {
+	rcic.mu.Lock()
+	rcic.seqs = append(rcic.seqs, &redisSequence{"SEND", commandName, args})
+	rcic.mu.Unlock()
```

Comment:
I am new to Go, so no action needed: Is there a preference between this and using `defer rcic.mu.Unlock()` after the `Lock()`? I see that pattern often but have not developed a heuristic around when to use it. Thanks!

Reply:
defer is slower (it pushes a function onto the stack). The performance has gotten a lot better in recent releases, but sometimes when you really care about performance, it's best to do this.